File: projects/chess-coach/apps/harness/src/main/kotlin/harness/Orchestrator.kt
// SPEC: _spec/chess-coach/harness/components.puml
// SPEC: _spec/chess-coach/harness/lifecycle.puml
package harness

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.Serializable

@Serializable
data class UiPhrases(val thinking: List<String>, val bestMove: List<String>)

class Orchestrator(
    private val stateManager: StateManager,
    private val engineBridge: EngineBridge,
    private val llmClient: LlmClient,
    private val skillLoader: SkillLoader = SkillLoader()
) {
    private val debugEnabled: Boolean =
        System.getenv("LLM_DEBUG")?.trim()?.equals("true", ignoreCase = true) == true

    suspend fun executeTurn(humanMove: String, difficulty: String = "intermediate"): TurnResult {
        System.err.println("--- Starting Turn ---")
        System.err.println("Human played: ${if (humanMove.isEmpty()) "(None - First Move)" else humanMove}")
        
        val currentFen = stateManager.readFen()
        
        System.err.println("Evaluating position...")
        val evalResult = engineBridge.getEvaluation(currentFen, depth = EngineConfig.Chess.DEFAULT_EVAL_DEPTH)
        val evalString = if (evalResult.isMate) "Mate in ${evalResult.mateIn}" else String.format("%.2f", evalResult.cp / 100.0)
        System.err.println("Stockfish Eval: Move: ${evalResult.bestMove}, Eval: $evalString")
        
        System.err.println("Asking Maia ($difficulty) for move...")
        val candidateMove = engineBridge.getMaiaMove(currentFen, difficulty)
        System.err.println("Maia Suggested Move: $candidateMove")
        
        val newFen = engineBridge.getFenAfterMove(currentFen, candidateMove) 
            ?: throw IllegalStateException("Maia generated an illegal move: $candidateMove")
        
        System.err.println("Move '$candidateMove' is LEGAL.")
        
        stateManager.writeFen(newFen)
        
        return TurnResult(
            fen = newFen,
            move = candidateMove,
            advice = ""
        )
    }

    suspend fun generateAdvice(humanMove: String, aiMove: String, currentFen: String): String {
        System.err.println("Generating coaching advice using ${llmClient.commentaryModel}...")
        val safeHumanMove = if (humanMove.isBlank()) "e4" else humanMove
        val evalBefore = engineBridge.getEvaluation(currentFen, depth = EngineConfig.Chess.DEFAULT_EVAL_DEPTH)
        val fenAfterMove = engineBridge.getFenAfterMove(currentFen, safeHumanMove)
        val evalAfter = if (fenAfterMove != null) {
            engineBridge.getEvaluation(fenAfterMove, depth = EngineConfig.Chess.DEFAULT_EVAL_DEPTH)
        } else {
            evalBefore
        }

        val analysis = createMoveAnalysis(
            fenBefore = currentFen,
            fenAfter = fenAfterMove ?: currentFen,
            moveSan = safeHumanMove,
            evalBefore = evalBefore,
            evalAfter = evalAfter,
            actor = EngineConfig.Llm.DEFAULT_ACTOR,
            gender = EngineConfig.Llm.DEFAULT_GENDER,
            mateScore = EngineConfig.Chess.MATE_SCORE_FOR_PROMPT,
            name = EngineConfig.Llm.DEFAULT_NAME
        )

        val adviceUserPrompt = buildPromptFromAnalysis(
            analysis = analysis,
            language = EngineConfig.Llm.DEFAULT_LANGUAGE,
            langCode = EngineConfig.Llm.DEFAULT_LANG_CODE,
            type = "standard"
        )
        
        val rawAdvice = llmClient.prompt(
            EngineConfig.Llm.SYSTEM_PROMPT,
            adviceUserPrompt,
            llmClient.commentaryModel,
            temperature = EngineConfig.Llm.DEFAULT_TEMPERATURE,
            maxTokens = EngineConfig.Llm.DEFAULT_MAX_TOKENS
        )
        System.err.println("Coach Advice generated.")

        val extractedCommentary = extractCommentary(rawAdvice)

        if (debugEnabled) {
            System.err.println(
                """
                |[LLM_DEBUG] kind=advice
                |prompt:
                |$adviceUserPrompt
                |analysis:
                |  moveSan=${analysis.moveSan}
                |  tag=${analysis.tag}
                |  bestAlt=${analysis.bestAlt}
                |  fenBefore=${analysis.fenBefore}
                |  fenAfter=${analysis.fenAfter}
                |raw:
                |$rawAdvice
                |extracted:
                |$extractedCommentary
                |final:
                |$extractedCommentary
                """.trimMargin()
            )
        }
        
        return extractedCommentary
    }

    suspend fun generateAdviceStream(humanMove: String, aiMove: String, currentFen: String): Flow<String> {
        System.err.println("Generating coaching advice stream using ${llmClient.commentaryModel}...")
        val safeHumanMove = if (humanMove.isBlank()) "e4" else humanMove
        val evalBefore = engineBridge.getEvaluation(currentFen, depth = EngineConfig.Chess.DEFAULT_EVAL_DEPTH)
        val fenAfterMove = engineBridge.getFenAfterMove(currentFen, safeHumanMove)
        val evalAfter = if (fenAfterMove != null) {
            engineBridge.getEvaluation(fenAfterMove, depth = EngineConfig.Chess.DEFAULT_EVAL_DEPTH)
        } else {
            evalBefore
        }

        val analysis = createMoveAnalysis(
            fenBefore = currentFen,
            fenAfter = fenAfterMove ?: currentFen,
            moveSan = safeHumanMove,
            evalBefore = evalBefore,
            evalAfter = evalAfter,
            actor = EngineConfig.Llm.DEFAULT_ACTOR,
            gender = EngineConfig.Llm.DEFAULT_GENDER,
            mateScore = EngineConfig.Chess.MATE_SCORE_FOR_PROMPT,
            name = EngineConfig.Llm.DEFAULT_NAME
        )

        val adviceUserPrompt = buildPromptFromAnalysis(
            analysis = analysis,
            language = EngineConfig.Llm.DEFAULT_LANGUAGE,
            langCode = EngineConfig.Llm.DEFAULT_LANG_CODE,
            type = "standard"
        )
        
        return collectCommentaryFlow(
            llmClient.promptStream(
                EngineConfig.Llm.SYSTEM_PROMPT,
                adviceUserPrompt,
                llmClient.commentaryModel,
                temperature = EngineConfig.Llm.DEFAULT_TEMPERATURE,
                maxTokens = EngineConfig.Llm.DEFAULT_MAX_TOKENS
            )
        )
    }

    suspend fun generateExplanation(
        fenBefore: String,
        fenAfter: String,
        moveSan: String,
        isBlunder: Boolean = true
    ): String {
        System.err.println("Generating explanation using ${llmClient.commentaryModel}...")
        val evalBefore = engineBridge.getEvaluation(fenBefore, depth = EngineConfig.Chess.DEFAULT_EVAL_DEPTH)
        val evalAfter = engineBridge.getEvaluation(fenAfter, depth = EngineConfig.Chess.DEFAULT_EVAL_DEPTH)

        val analysis = createMoveAnalysis(
            fenBefore = fenBefore,
            fenAfter = fenAfter,
            moveSan = moveSan,
            evalBefore = evalBefore,
            evalAfter = evalAfter,
            actor = EngineConfig.Llm.DEFAULT_ACTOR,
            gender = EngineConfig.Llm.DEFAULT_GENDER,
            mateScore = EngineConfig.Chess.MATE_SCORE_FOR_PROMPT,
            isForcedBlunder = isBlunder,
            name = EngineConfig.Llm.DEFAULT_NAME
        )

        val explainUserPrompt = buildPromptFromAnalysis(
            analysis = analysis,
            language = EngineConfig.Llm.DEFAULT_LANGUAGE,
            langCode = EngineConfig.Llm.DEFAULT_LANG_CODE,
            type = "explanation"
        )
        
        val rawExplanation = llmClient.prompt(
            EngineConfig.Llm.SYSTEM_PROMPT,
            explainUserPrompt,
            llmClient.commentaryModel,
            temperature = EngineConfig.Llm.EXPLANATION_TEMPERATURE,
            maxTokens = EngineConfig.Llm.EXPLANATION_MAX_TOKENS
        )
        System.err.println("Explanation generated.")

        val extractedCommentary = extractCommentary(rawExplanation)

        if (debugEnabled) {
            System.err.println(
                """
                |[LLM_DEBUG] kind=explanation
                |prompt:
                |$explainUserPrompt
                |analysis:
                |  moveSan=${analysis.moveSan}
                |  tag=${analysis.tag}
                |  bestAlt=${analysis.bestAlt}
                |  fenBefore=${analysis.fenBefore}
                |  fenAfter=${analysis.fenAfter}
                |raw:
                |$rawExplanation
                |extracted:
                |$extractedCommentary
                |final:
                |$extractedCommentary
                """.trimMargin()
            )
        }
        
        return extractedCommentary
    }

    suspend fun generateExplanationStream(fenBefore: String, fenAfter: String, isBlunder: Boolean, moveSan: String): Flow<String> {
        System.err.println("Generating explanation stream (isBlunder=$isBlunder, move=$moveSan) using ${llmClient.commentaryModel}...")
        val evalBefore = engineBridge.getEvaluation(fenBefore, depth = EngineConfig.Chess.DEFAULT_EVAL_DEPTH)
        val evalAfter = engineBridge.getEvaluation(fenAfter, depth = EngineConfig.Chess.DEFAULT_EVAL_DEPTH)

        val analysis = createMoveAnalysis(
            fenBefore = fenBefore,
            fenAfter = fenAfter,
            moveSan = moveSan,
            evalBefore = evalBefore,
            evalAfter = evalAfter,
            actor = EngineConfig.Llm.DEFAULT_ACTOR,
            gender = EngineConfig.Llm.DEFAULT_GENDER,
            mateScore = EngineConfig.Chess.MATE_SCORE_FOR_PROMPT,
            isForcedBlunder = isBlunder,
            name = EngineConfig.Llm.DEFAULT_NAME
        )

        val explainUserPrompt = buildPromptFromAnalysis(
            analysis = analysis,
            language = EngineConfig.Llm.DEFAULT_LANGUAGE,
            langCode = EngineConfig.Llm.DEFAULT_LANG_CODE,
            type = "explanation"
        )
        
        return collectCommentaryFlow(
            llmClient.promptStream(
                EngineConfig.Llm.SYSTEM_PROMPT,
                explainUserPrompt,
                llmClient.commentaryModel,
                temperature = EngineConfig.Llm.EXPLANATION_TEMPERATURE,
                maxTokens = EngineConfig.Llm.EXPLANATION_MAX_TOKENS
            )
        )
    }

    private suspend fun collectCommentaryFlow(source: Flow<String>): Flow<String> = flow {
        val buffer = StringBuilder()
        source.collect { chunk ->
            buffer.append(chunk)
        }
        val commentary = extractCommentary(buffer.toString())
        if (commentary.isNotBlank()) {
            emit(commentary)
        }
    }

    suspend fun generateUiPhrases(): UiPhrases {
        try {
            System.err.println("Warming up LLM into VRAM...")
            llmClient.prompt("System", "Ping", llmClient.commentaryModel, maxTokens = 1)
            System.err.println("LLM Warmup complete.")
        } catch (e: Exception) {
            System.err.println("LLM Warmup failed: ${e.message}")
        }

        return UiPhrases(
            thinking = listOf("Hmm...", "Let me think...", "Interesting position...", "Rats...", "What to do..."),
            bestMove = listOf("Great move!", "Excellent!", "I like that.", "Strong play.", "Well done.")
        )
    }
}

data class TurnResult(
    val fen: String,
    val move: String,
    val advice: String
)

File: projects/chess-coach/apps/server/src/Orchestrator.ts
import { buildPromptFromAnalysis, createMoveAnalysis, extractCommentary } from "@chess-coach/engine-core";

import { EngineConfig } from "./config.ts";
import type { IEngineBridge } from "./EngineBridge.ts";
import type { LlmClient } from "./LlmClient.ts";
import type { StateManager } from "./StateManager.ts";

export type TurnResult = { fen: string; move: string; advice: string };
export type UiPhrases = { thinking: string[]; bestMove: string[] };

const DEBUG_ENABLED = (process.env.LLM_DEBUG ?? "").trim().toLowerCase() === "true";

/**
 * In-process orchestrator. Faithful port of the Kotlin harness Orchestrator
 * (_spec/api/behavior.md §4), built on the shared @chess-coach/engine-core
 * prompt logic. There is no daemon / JSON-line layer — the HTTP handler calls
 * these methods directly.
 */
export class Orchestrator {
  constructor(
    private readonly stateManager: StateManager,
    private readonly engineBridge: IEngineBridge,
    private readonly llmClient: LlmClient,
  ) {}

  async executeTurn(humanMove: string, difficulty = "intermediate"): Promise<TurnResult> {
    console.error("--- Starting Turn ---");
    console.error(`Human played: ${humanMove === "" ? "(None - First Move)" : humanMove}`);

    const currentFen = this.stateManager.readFen();

    console.error("Evaluating position...");
    const evalResult = await this.engineBridge.getEvaluation(
      currentFen,
      EngineConfig.chess.defaultEvalDepth,
    );
    const evalString = evalResult.isMate
      ? `Mate in ${evalResult.mateIn}`
      : (evalResult.cp / 100).toFixed(2);
    console.error(`Stockfish Eval: Move: ${evalResult.bestMove}, Eval: ${evalString}`);

    console.error(`Asking Maia (${difficulty}) for move...`);
    const candidateMove = await this.engineBridge.getMaiaMove(currentFen, difficulty);
    console.error(`Maia Suggested Move: ${candidateMove}`);

    const newFen = await this.engineBridge.getFenAfterMove(currentFen, candidateMove);
    if (!newFen) throw new Error(`Maia generated an illegal move: ${candidateMove}`);

    console.error(`Move '${candidateMove}' is LEGAL.`);
    this.stateManager.writeFen(newFen);

    return { fen: newFen, move: candidateMove, advice: "" };
  }

  async *generateAdviceStream(
    humanMove: string,
    _aiMove: string,
    currentFen: string,
  ): AsyncGenerator<string, void, unknown> {
    const safeHumanMove = humanMove.trim() === "" ? "e4" : humanMove;

    const evalBefore = await this.engineBridge.getEvaluation(
      currentFen,
      EngineConfig.chess.defaultEvalDepth,
    );
    const fenAfterMove = await this.engineBridge.getFenAfterMove(currentFen, safeHumanMove);
    const evalAfter = fenAfterMove
      ? await this.engineBridge.getEvaluation(fenAfterMove, EngineConfig.chess.defaultEvalDepth)
      : evalBefore;

    const analysis = createMoveAnalysis({
      fenBefore: currentFen,
      fenAfter: fenAfterMove ?? currentFen,
      moveSan: safeHumanMove,
      evalBefore,
      evalAfter,
      actor: EngineConfig.llm.defaults.actor,
      gender: EngineConfig.llm.defaults.gender,
      mateScore: EngineConfig.chess.mateScoreForPrompt,
      name: EngineConfig.llm.defaults.name,
    });

    const prompt = buildPromptFromAnalysis(
      analysis,
      EngineConfig.llm.defaults.language,
      EngineConfig.llm.defaults.langCode,
      "standard",
    );

    let raw = "";
    for await (const chunk of this.llmClient.promptStream(
      EngineConfig.llm.systemPrompt,
      prompt,
      this.llmClient.commentaryModel,
      EngineConfig.llm.defaultTemperature,
      EngineConfig.llm.defaultMaxTokens,
    )) {
      raw += chunk;
    }

    const commentary = extractCommentary(raw);
    if (DEBUG_ENABLED) {
      console.error(`[LLM_DEBUG] kind=advice\nprompt:\n${prompt}\nraw:\n${raw}\nextracted:\n${commentary}`);
    }
    if (commentary.trim() !== "") yield commentary;
  }

  async *generateExplanationStream(
    fenBefore: string,
    fenAfter: string,
    isBlunder: boolean,
    moveSan: string,
  ): AsyncGenerator<string, void, unknown> {
    const evalBefore = await this.engineBridge.getEvaluation(
      fenBefore,
      EngineConfig.chess.defaultEvalDepth,
    );
    const evalAfter = await this.engineBridge.getEvaluation(
      fenAfter,
      EngineConfig.chess.defaultEvalDepth,
    );

    const analysis = createMoveAnalysis({
      fenBefore,
      fenAfter,
      moveSan,
      evalBefore,
      evalAfter,
      actor: EngineConfig.llm.defaults.actor,
      gender: EngineConfig.llm.defaults.gender,
      mateScore: EngineConfig.chess.mateScoreForPrompt,
      isForcedBlunder: isBlunder,
      name: EngineConfig.llm.defaults.name,
    });

    const prompt = buildPromptFromAnalysis(
      analysis,
      EngineConfig.llm.defaults.language,
      EngineConfig.llm.defaults.langCode,
      "explanation",
    );

    let raw = "";
    for await (const chunk of this.llmClient.promptStream(
      EngineConfig.llm.systemPrompt,
      prompt,
      this.llmClient.commentaryModel,
      EngineConfig.llm.explanationTemperature,
      EngineConfig.llm.explanationMaxTokens,
    )) {
      raw += chunk;
    }

    const commentary = extractCommentary(raw);
    if (DEBUG_ENABLED) {
      console.error(`[LLM_DEBUG] kind=explanation\nprompt:\n${prompt}\nraw:\n${raw}\nextracted:\n${commentary}`);
    }
    if (commentary.trim() !== "") yield commentary;
  }

  async generateUiPhrases(): Promise<UiPhrases> {
    try {
      console.error("Warming up LLM into VRAM...");
      await this.llmClient.prompt(
        "System",
        "Ping",
        this.llmClient.commentaryModel,
        EngineConfig.llm.defaultTemperature,
        1,
      );
      console.error("LLM Warmup complete.");
    } catch (e) {
      console.error(`LLM Warmup failed: ${(e as Error).message}`);
    }

    return {
      thinking: ["Hmm...", "Let me think...", "Interesting position...", "Rats...", "What to do..."],
      bestMove: ["Great move!", "Excellent!", "I like that.", "Strong play.", "Well done."],
    };
  }
}
